keps/sig-scheduling/4832-async-preemption/README.md

```diff
@@ -0,0 +1,999 @@
+
+So, we don't have to pay a special attention to this issue.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+To achieve an asynchronous preemption, we will change the preemption plugin's implementation like the following:
+1. The preemption PostFilter plugin calculates the preemption target and nominate the Pod for the Node. (We'll use `AddNominatedPod` API exposed from the scheduling framework to plugins.)
+2. The preemption PostFilter plugin starts the goroutine to make API calls inside, and return success status (= not wait for the goroutine to finish).
+3. The preemption plugin gates the Pod, which the preemption is in-progress, at PreEnqueue extension point so that the target Pod won't be retried during the preemption.
+
+Then, afterwards the preemption goroutine makes actual API calls. 
+If the preemption goroutine fails at some point, it reverts the nomination via `AddNominatedPod` with [`clearNominatedNode`](https://github.com/kubernetes/kubernetes/blob/f5c538418189e119a8dbb60e2a2b22394548e326/pkg/scheduler/schedule_one.go#L135).
+If the preemption goroutine succeeds, the Pod is queued back to the queue, and (hopefully) scheduled on the nominated node.
+
+### Consideration to race condition
+
+Thanks to the nomination at `PostFilter`, this new asynchronous preemption shouldn't make any race condition between several scheduling cycles.
+
+Here, I'll discuss what happens in which scenario, and make sure there's no worry.
+
+Let's say pod1 is during the preemption process (node1) at the preemption goroutine, the next scheduling cycle is scheduling pod2. 
+
+#### The pod2's scheduling is successful (pod2 is equal or lower priority than pod1)
+
+As I described above, pod1's `PostFilter` nominates pod1 for node1.
+
+At the scheduling cycle, the scheduler takes such nominated pods that are equal or higher priority than pod1 into consideration;
+meaning, pod2 won't rob pod1 of the place on node1.
+
+#### The pod2's scheduling is successful (pod2 is higher priority than pod1)
+
+Even though pod1 is nominated for the node, the scheduler allows pod2 to take node1, where the pod1's preemption made the space. 
+
+Then, when pod1 comes back to the scheduling cycle, it may not be able to land on node1 because pod2 is scheduled there now.
+It happens with both the current and this KEP's scheduler, so no issue here.
+
+#### The pod2's scheduling is failed and starts the preemption (pod2 is equal or lower priority than pod1)
+
+The preemption also takes nominated Pods into consideration when calculating the preemption target.
+
+Therefore, if, coincidently, two preemptions for pod1 and pod2 select the same Node after all, 
```

Comment:
Could pod2 issue preemptions for pods that are already preempted or being preempted in a routine?

I found this comment https://github.com/kubernetes/kubernetes/blob/851cf43a35862aeeafc4a0966f7d1e0836f675e4/pkg/scheduler/framework/runtime/framework.go#L969-L972, but I'm not sure where the implementation of that is (removing victims)

Reply:
> Could pod2 issue preemptions for pods that are already preempted or being preempted in a routine?

Yes. So, when we delete Pods, we have to ignore not-found errors from the API.

So, let's say node1 has `victim-pod1` - `victim-pod4` with priority 1 - 4 (correspondingly), and each of them requires two CPUs.
Let's say `preemptor-pod1` and `preemptor-pod2` have the same priority that is higher than all Pods on node1, and both of them require 4 CPUs.
In this case, what is supposed to happen is that
1. `preemptor-pod1` selects `victim-pod1` and `victim-pod2` for the preemption victims because they have lower priorities than other Pods on node1.
2. (let's say the preemption API calls for pod1 haven't finished at all yet) `preemptor-pod2` selects `victim-pod1`, `victim-pod2`, `victim-pod3` and `victim-pod4` for the preemption target.
This is because, during the target calculation, the scheduler assumes `preemptor-pod1` is already on the node1; meaning, node1 has more Pods than its actual capacity. So, if `preemptor-pod2` selects node1 for the preemption, it has to delete enough Pods to make 8 CPUs.

---

> I found this comment

I think... the comment is invalid? I couldn't find such implementation either.
So, I guess actually, for the same reason described above, the current implementation goes without any issue, without deleting victims in `SelectVictimsOnNode`.

Reply:
Ok, so `preemptor-pod2` would still choose `victim-pod1` and `victim-pod2`. That is fine, however, we could save API calls if these pods already have a `deletionTimestamp`. This is an implementation detail, but please keep it in mind.